### PR TITLE
Generic Data Structures

### PR DESCRIPTION
**Pull request by elazarg**
Decompose the data structures into domain-specific and generic data structures.
 
The `ds` package has zero dependencies, and it knows nothing about numbers or network.

Interfaces (all in `interfaces.go`):
* Comparable (Equal, Copy)
* Hashable (Comparable, Hash)
* Set (Hashable + set operations)
* Product of A x B (partitions, left and right projections, swap)
* TripleSet, convenience (and associativity agnostic) interface for A x B x C

Implementations:
* `HashMap` - a simple adapter around a go map, allowing `Hashable` keys.
* `HashSet`
* `MultiMap` is a non-injective map. It gives a simple "inverse" on a HashMap.
* `ProductLeft` <s>and ProductRight</s> is a map-based implementation of a cartesian product; it implements `Product` as a one-to-one mapping from sets to sets, where each key-value pairs encodes the cross product of its two items.
* `TripleSetLeft` and `TripleSetRight` and `TripleSetOuter` are right-, left- and outer-associative implementations of `TripleSet`.
* `DisjointSum` is tagged-union of two sets. Go's generics place some limitations on the implementation; The empty element must always be explicitly passed to the constructor.
* <s>`UnitSet` is an idea - using sets of size 1 as the identity element of Product. It's probably not really helpful.</s>

The `interval` package has `interval` and `intervalset`. Nothing particularly new about them.

`netp` is unchanged.

The `netset` package is where handles sets of network elements.
* `ICMPSet` is largely the one proposed in #23.
* `IPBlock` is adapted to implement `Set`.
* `TCPUDPSet` is a `TripleSet` of  `ProtocolSet` x `PortSet` x `PortSet`
* `TransportSet` is a disjoint union of `TCPUDPSet` + `ICMPSet`
* `ConnectionSet` is  a `TripleSet` of `IPBlock` x `IPBlock` x `TransportSet`

The package `connection` now contains the state-aware connectionset (named `Set`), and its json-formatting. In my opinion it should be moved to the analyzer.

Many tests are adapted. <s>but some data structures do not have dedicated tests.</s>

There is no implementation of a dynamically-bounded tuple, since I could not find the proper way to fit that into a generic datastructure with a well defined degenerate edge case.

The diff is probably not useful; it should be easier to read the code.

### CODE REVIEW

File: pkg/netset/connectionset.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+// Subtract
+// ToDo: Subtract seems to ignore IsStateful (see https://github.com/np-guard/vpc-network-config-analyzer/issues/199):
+//  1. is the delta connection stateful
+//  2. props is identical but c stateful while other is not
+//     the 2nd item can be computed here, with enhancement to relevant structure
+//     the 1st can not since we do not know where exactly the statefulness came from
+func (c *ConnectionSet) Subtract(other *ConnectionSet) *ConnectionSet {
+	if c.IsEmpty() {
+		return &ConnectionSet{props: ds.NewRightTripleSet[*IPBlock, *IPBlock, *ds.Disjoint[*TCPUDPSet, *ICMPSet]]()}
```
Comment (adisos) [inline pkg/netset/connectionset.go:69]:
```suggestion
		return &ConnectionSet{props: ds.NewRightTripleSet[*IPBlock, *IPBlock, *TransportSet]()}
```

File: pkg/netset/tcpudpset.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	protocol := protocolStringToCode(protocolString)
+	return path(
+		interval.New(protocol, protocol).ToSet(),
+		interval.New(srcMinP, srcMaxP).ToSet(),
+		interval.New(dstMinP, dstMaxP).ToSet(),
+	)
+}
+
+func EmptyTCPorUDPSet() *TCPUDPSet {
+	return &TCPUDPSet{props: ds.NewRightTripleSet[*PortSet, *PortSet, *PortSet]()}
```
Comment (adisos) [inline pkg/netset/tcpudpset.go:172]:
```suggestion
	return &TCPUDPSet{props: ds.NewRightTripleSet[*ProtocolSet, *PortSet, *PortSet]()}
```

File: pkg/ds/interfaces.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	Hashable[Self]
+	IsEmpty() bool
+	Size() int
+	IsSubset(Self) bool
+	Union(Self) Self
+	Intersect(Self) Self
+	Subtract(Self) Self
+}
+
+// Product is a subset of cartesian product of sets S1 x S2
```
Comment (adisos) [inline pkg/ds/interfaces.go:26]:
```suggestion
// Product is a cartesian product of sets S1 x S2
```

File: pkg/ds/hashmap.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			return
+		}
+	}
+	m.m[k.Hash()] = append(Pairs, Pair[K, V]{Left: k.Copy(), Right: v.Copy()})
+}
+
+func (m *HashMap[K, V]) Copy() *HashMap[K, V] {
+	res := NewMap[K, V]()
+	for _, p := range m.Pairs() {
+		res.Insert(p.Left.Copy(), p.Right.Copy())
```
Comment (adisos) [inline pkg/ds/hashmap.go:48]: `Insert` already copies.. could be changed to `res.Insert(p.Left, p.Right)` ?

File: pkg/ds/hashmap.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+func (m *HashMap[K, V]) Keys() []K {
+	pairs := m.Pairs()
+	res := make([]K, len(pairs))
+	for i, p := range m.Pairs() {
+		res[i] = p.Left
+	}
+	return res
+}
+
+func (m *HashMap[K, V]) Values() []V {
```
Comment (adisos) [inline pkg/ds/hashmap.go:80]: This function may return a list with duplicated values?

File: pkg/netset/connectionset.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func NewICMPTransport(tc netp.ICMP) *TransportSet {
+	return ds.NewDisjoint(
+		EmptyTCPorUDPSet(),
+		NewICMPSet(tc),
+	)
+}
+
+type ConnectionSet struct {
```
Comment (adisos) [inline pkg/netset/connectionset.go:27]: missing `NewConnectionSet()` function?

**Review by adisos — COMMENTED** [review]
Looks good, added few initial comments.

File: pkg/ds/hashmap.go (hunk 5 above)
Comment (elazarg) [reply pkg/ds/hashmap.go:80]: Yes, why not?

File: pkg/netset/connectionset.go (hunk 6 above)
Comment (elazarg) [reply pkg/netset/connectionset.go:27]: added

File: pkg/ds/interfaces.go (hunk 3 above)
Comment (elazarg) [reply pkg/ds/interfaces.go]: Changed

File: pkg/netset/tcpudpset.go (hunk 2 above)
Comment (elazarg) [reply pkg/netset/tcpudpset.go]: Done

File: pkg/netset/connectionset.go (hunk 1 above)
Comment (elazarg) [reply pkg/netset/connectionset.go]: Done

**Comment by elazarg** [comment]
We may need to use ordered maps instead of hash maps, so the lists of partitions are consistently ordered.

File: pkg/ds/interfaces.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	Intersect(Self) Self
+	Subtract(Self) Self
+}
+
+// Product is a cartesian product of sets S1 x S2
+type Product[S1 Set[S1], S2 Set[S2]] interface {
+	Set[Product[S1, S2]]
+	Partitions() []Pair[S1, S2]
+	NumPartitions() int
+	Left() []S1
```
Comment (adisos) [inline pkg/ds/interfaces.go:44]: shouldn't `Left()` return `Set[S1]` instead of `[]S1` ?
Shouldn't the projection on one dimension be just a set without order and with no duplicated elements?

File: pkg/ds/interfaces.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+// designed to hold cross-products of two Set structures in a succinct and canonical way.
+// The most important interface is Product. It is implemented by ProductLeft using injective
+// (one-to-one) mapping from sets to sets, where each key-value pair defines a complete cross
+// product of the two sets.
+package ds
+
+type Sized interface {
+	IsEmpty() bool
+	// Size returns the actual, full size of the set. For Product, it returns the number of pairs.
+	Size() int
```
Comment (adisos) [inline pkg/ds/interfaces.go:17]: should `IsAll()` be also part of the interface?

File: pkg/ds/interfaces.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	Equal(Self) bool
+	Copy() Self
+}
+
+type Hashable[Self any] interface {
+	Comparable[Self]
+	Hash() int
+}
+
+type Set[Self any] interface {
```
Comment (adisos) [inline pkg/ds/interfaces.go:30]: should `String()` be also part of the interfaces of sets?

File: pkg/ds/interfaces.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	IsSubset(Self) bool
+	Union(Self) Self
+	Intersect(Self) Self
+	Subtract(Self) Self
+}
+
+// Product is a cartesian product of sets S1 x S2
+type Product[S1 Set[S1], S2 Set[S2]] interface {
+	Set[Product[S1, S2]]
+	Partitions() []Pair[S1, S2]
```
Comment (adisos) [inline pkg/ds/interfaces.go:42]: Should `Partitions()` return `Set[Pair[S1,S2]]` ? (order should not matter?)

File: pkg/netset/icmpset.go
Comment (adisos) [inline pkg/netset/icmpset.go:1]: need some documentation of the various vars and functions in this file

File: pkg/netset/connectionset.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+		props: c.props.Union(other.props),
+	}
+}
+
+// Subtract
+// ToDo: Subtract seems to ignore IsStateful (see https://github.com/np-guard/vpc-network-config-analyzer/issues/199):
+//  1. is the delta connection stateful
+//  2. props is identical but c stateful while other is not
+//     the 2nd item can be computed here, with enhancement to relevant structure
+//     the 1st can not since we do not know where exactly the statefulness came from
```
Comment (adisos) [inline pkg/netset/connectionset.go:56]:
```suggestion
```

File: pkg/ds/unitset.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package ds
+
+type UnitSet struct {
```
Comment (adisos) [inline pkg/ds/unitset.go:9]: `UnitSet` not used?

File: pkg/ds/tripleset_outer.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package ds
+
+// OuterTripleSet is an outer-associative 3-product of sets (S1 x S3) x S2
+type OuterTripleSet[S1 Set[S1], S2 Set[S2], S3 Set[S3]] struct {
```
Comment (adisos) [inline pkg/ds/tripleset_outer.go:10]: why do we need `OuterTripleSet` and `RightTripleSet` ?

File: pkg/netp/icmp.go
Comment (adisos) [inline pkg/netp/icmp.go:1]: Also need some documentation here

**Review by adisos — COMMENTED** [review]
few more questions and comments

File: pkg/ds/interfaces.go (hunk 8 above)
Comment (elazarg) [reply pkg/ds/interfaces.go:17]: I did not think it is a general property; it requires knowledge about the "universe" which sets may not have.

File: pkg/ds/interfaces.go (hunk 10 above)
Comment (elazarg) [reply pkg/ds/interfaces.go:42]: You are probably right.

File: pkg/ds/tripleset_outer.go (hunk 13 above)
Comment (elazarg) [reply pkg/ds/tripleset_outer.go:10]: It implies different partitions.

File: pkg/ds/unitset.go (hunk 12 above)
Comment (elazarg) [reply pkg/ds/unitset.go:9]: No. I thought it could be used as the edge case of a dynamically-bounded product, but it didn't work.

File: pkg/ds/interfaces.go (hunk 9 above)
Comment (elazarg) [reply pkg/ds/interfaces.go:30]: It is not required for the data structure, and would need decisions that I thought are best left for client code.

File: pkg/ds/interfaces.go (hunk 7 above)
Comment (elazarg) [reply pkg/ds/interfaces.go:44]: This sounds correct, however, there is no obvious way to return an empty set for the empty partition. It may need an argument to Left() and Right() just to handle this case.

**Comment by elazarg** [comment]
Added an elaborated README, and changed implementation of MultiMap to use HashSet.

**Comment by elazarg** [comment]
BTW, I think the class ICMP is better named ICMPRule, and is a set. The file icmp.go in the netp package should only have functions that are related to ICMP proper. Such change would also solve the need to expose implementation details of the ICMP class to the ICMPSet class.

**Comment by elazarg** [comment]
@adisos is there anything left for me to implement/document so this PR can be merged?

File: pkg/ds/hashmap.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package ds
+
+// HashMap is a generic hash map with keys of a Hashable type K and values of Comparable type V.
+type HashMap[K Hashable[K], V Comparable[V]] struct {
+	m map[int][]Pair[K, V]
```
Comment (adisos) [inline pkg/ds/hashmap.go:11]:
```suggestion
	m map[int][]Pair[K, V] // map from key hash value to the list of pairs (k,v)
```

File: pkg/ds/product_left.go [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+	}
+	return res
+}
+
+// IsEmpty returns true if the Product object is empty.
+func (m *ProductLeft[K, V]) IsEmpty() bool {
+	return m.m.IsEmpty()
+}
+
+// Size returns the number of unique pairs in the Product object
```
Comment (adisos) [inline pkg/ds/product_left.go:88]: the Size documentation is incorrect? it does not return the number of pairs..

File: pkg/ds/interfaces.go (hunk 10 above)
Comment (adisos) [inline pkg/ds/interfaces.go:42]:
```suggestion
        // Partitions returns a slice of pairs in the product set  (note that the order is arbitrary)
	Partitions() []Pair[S1, S2]
```

File: pkg/ds/interfaces.go [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+	NumPartitions() int
```
Comment (adisos) [inline pkg/ds/interfaces.go:43]: should explain the difference between `NumPartitions` and `Size` ?

File: pkg/ds/interfaces.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+	Left(empty S1) S1
```
Comment (adisos) [inline pkg/ds/interfaces.go:44]:
```suggestion
        // Left returns the left projection from pairs in the product set on S1, given input of an empty set in S1
	Left(empty S1) S1
```

File: pkg/ds/interfaces.go [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+	Right(empty S2) S2
```
Comment (adisos) [inline pkg/ds/interfaces.go:45]:
```suggestion
        // Right returns the right projection from pairs in the product set S2, given input of an empty set in S2
	Right(empty S2) S2
```

File: pkg/ds/interfaces.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+	Swap() Product[S2, S1]
```
Comment (adisos) [inline pkg/ds/interfaces.go:46]:
```suggestion
        // Swap returns a new Product object, built from the input object, with left and right swapped
	Swap() Product[S2, S1]
```

**Review by adisos — COMMENTED** [review]
few small comments and questions

**Comment by adisos** [comment]
> @adisos is there anything left for me to implement/document so this PR can be merged?

added few small comments, still reviewing the tests and some implementation details.

**Comment by elazarg** [comment]
interval:
* Separate tests from interval_test.go
* Improve documentation.
* Export and set-like functions that are well defined.
* Rename interval.Subtract to interval.SubtractSplit, and add tests.
* Handle empty cases first.
* Preallocate Elements.

intervalset:
* Guard Size() from overflow, and use intervalset.CalculateSize().
* Handle empty cases first.
* Remove String() method, since it is not obvious; clients should implement.

---

I think it's better to make `GenericInterval` and `GenericCanonicalSet` as generic types (with integral constraints), so clients can choose whether or not to rely on things like `Size()` being 64bit integer. `interval.Interval` should be an alias to `GenericInterval[int]`, and interval.Set should be an alias to `GenericCanonicalSet[int]`.

File: pkg/ds/interfaces.go (hunk 10 above)
Comment (elazarg) [reply pkg/ds/interfaces.go:49]: This is incorrect. It returns a slice of pairs, where the union of their cross product contains exactly the same pairs as this `Product`.

File: pkg/ds/interfaces.go (hunk 16 above)
Comment (elazarg) [reply pkg/ds/interfaces.go:52]: I don't understand this comment. The doc here explains exactly that:
> NumPartitions returns len(Partitions()). It is different from Size() which should return the number of concrete pairs of elements.

File: pkg/ds/product_left.go (hunk 15 above)
Comment (elazarg) [reply pkg/ds/product_left.go:88]: I'm not sure I understand. Maybe the issue here is the notion of "pair" - here it means the mathematical pair of underlying elements, [x, y], not a pair of sets [s1, s2] (usually implicitly encoding s1xs2, as is the case in the return value of `Partitions()`.

File: pkg/ds/interfaces.go (hunk 10 above)
Comment (elazarg) [reply pkg/ds/interfaces.go:50]: Docs are updated:
```golang
// Partitions returns a slice of pairs such that, for (p Product):
// 	  p.Equal(Union(CartesianPairLeft(s1, s2) for _, (s1, s2) := range p.Partitions())
// (note that the order is arbitrary; we do not return HashSet because Pair is not Hashable)
// Partitions returns a slice of pairs in the product set  (note that the order is arbitrary)
```

**Comment by YairSlobodin1** [comment]
Hi @elazarg,
Could you please add [IPBlock.Overlap](https://github.com/np-guard/models/blob/main/pkg/ipblock/ipblock.go#L119) method to the code?

**Comment by elazarg** [comment]
> Hi @elazarg, Could you please add the [IPBlock.Overlap](https://github.com/np-guard/models/blob/main/pkg/ipblock/ipblock.go#L119) method to the code?

@YairSlobodin1 Done

**Comment by YairSlobodin1** [comment]
Hi @elazarg,
Thanks for adding the `IPBlock.Overlap` method to the code!
Could you please also add the `NewTCPUDP` function, following the same idea as [ICMPFromTypeAndCode](https://github.com/np-guard/models/blob/generic/pkg/netp/icmp.go#L44)?

**Comment by elazarg** [comment]
@YairSlobodin1 not sure what exactly you are asking for. Can you suggest an implementation?

**Comment by YairSlobodin1** [comment]
@elazarg this is what I had in mind, but please feel free to adjust as needed.


```
func NewTCPUDP(isTCP bool, minSrcPort, maxSrcPort, minDstPort, maxDstPort int) (TCPUDP, error) {
	if min(minSrcPort, minDstPort) < MinPort || max(maxSrcPort, maxDstPort) > MaxPort {
		return TCPUDP{}, fmt.Errorf("TCPUDP ports are in range %d-%d", MinPort, MaxPort)
	} 
	srcPorts := interval.New(int64(minSrcPort), int64(maxSrcPort))
	dstPorts := interval.New(int64(minDstPort), int64(maxDstPort))
	return TCPUDP{IsTCP: isTCP, PortRangePair: PortRangePair{SrcPort: srcPorts, DstPort: dstPorts}}, nil
}
```

**Comment by YairSlobodin1** [comment]
@elazarg this is what I had in mind, but feel free to adjust as needed.


```
func NewTCPUDP(isTCP bool, minSrcPort, maxSrcPort, minDstPort, maxDstPort int) (TCPUDP, error) {
	if min(minSrcPort, minDstPort) < MinPort || max(maxSrcPort, maxDstPort) > MaxPort {
		return TCPUDP{}, fmt.Errorf("TCPUDP ports are in range %d-%d", MinPort, MaxPort)
	} 
	srcPorts := interval.New(int64(minSrcPort), int64(maxSrcPort))
	dstPorts := interval.New(int64(minDstPort), int64(maxDstPort))
	return TCPUDP{IsTCP: isTCP, PortRangePair: PortRangePair{SrcPort: srcPorts, DstPort: dstPorts}}, nil
}
```

**Comment by YairSlobodin1** [comment]
Hi @elazarg,
Could you please also add the `IPBlock.IsIPAddress()` method? 
```
func (b *IPBlock) IsIPAddress() bool {
	return b.Size() == 1
}
```

**Comment by elazarg** [comment]
@YairSlobodin1 there's already ToIPAddressString() which returns "" for non-singleton blocks. What is your use case?

> Hi @elazarg, Could you please also add the `IPBlock.IsIPAddress()` method?
> 
> ```
> func (b *IPBlock) IsIPAddress() bool {
> 	return b.Size() == 1
> }
> ```

**Comment by YairSlobodin1** [comment]
> @YairSlobodin1 there's already ToIPAddressString() which returns "" for non-singleton blocks. What is your use case?
> 
> > Hi @elazarg, Could you please also add the `IPBlock.IsIPAddress()` method?
> > ```
> > func (b *IPBlock) IsIPAddress() bool {
> > 	return b.Size() == 1
> > }
> > ```

@elazarg I have some `if` statements in synthesis that check if the `IPBlock` is a single IP address

**Comment by elazarg** [comment]
> > @YairSlobodin1 there's already ToIPAddressString() which returns "" for non-singleton blocks. What is your use case?
> > > Hi @elazarg, Could you please also add the `IPBlock.IsIPAddress()` method?
> > > ```
> > > func (b *IPBlock) IsIPAddress() bool {
> > > 	return b.Size() == 1
> > > }
> > > ```
> 
> @elazarg I have some `if` statements in synthesis that check if the `IPBlock` is a single IP address

And what do you do then?
It is possible to use ToIPAddressString() for this check. Alternatively, we may need to have some other API, depending on the usage.

**Comment by YairSlobodin1** [comment]
> > > @YairSlobodin1 there's already ToIPAddressString() which returns "" for non-singleton blocks. What is your use case?
> > > > Hi @elazarg, Could you please also add the `IPBlock.IsIPAddress()` method?
> > > > ```
> > > > func (b *IPBlock) IsIPAddress() bool {
> > > > 	return b.Size() == 1
> > > > }
> > > > ```
> > 
> > 
> > @elazarg I have some `if` statements in synthesis that check if the `IPBlock` is a single IP address
> 
> And what do you do then? It is possible to use ToIPAddressString() for this check. Alternatively, we may need to have some other API, depending on the usage.

[example](https://github.com/np-guard/vpc-network-config-synthesis/blob/use_models/pkg/io/confio/sg.go#L56)

**Comment by elazarg** [comment]
> > > > @YairSlobodin1 there's already ToIPAddressString() which returns "" for non-singleton blocks. What is your use case?
> > > > > Hi @elazarg, Could you please also add the `IPBlock.IsIPAddress()` method?
> > > > > ```
> > > > > func (b *IPBlock) IsIPAddress() bool {
> > > > > 	return b.Size() == 1
> > > > > }
> > > > > ```
> > > 
> > > 
> > > @elazarg I have some `if` statements in synthesis that check if the `IPBlock` is a single IP address
> > 
> > 
> > And what do you do then? It is possible to use ToIPAddressString() for this check. Alternatively, we may need to have some other API, depending on the usage.
> 
> [example](https://github.com/np-guard/vpc-network-config-synthesis/blob/use_models/pkg/io/confio/sg.go#L56)

I think the following is better, since you are guaranteed that the format is actualy IP format and not some encoding of size-1 range:
```golang
	if ipString := t.ToIPAddressString(); ipString != "" {
		return &vpcv1.SecurityGroupRuleRemoteIP{
			Address: &ipString,
		}
	}
	return &vpcv1.SecurityGroupRuleRemoteCIDR{
		CIDRBlock: &st,
	}
```
(Maybe it's better to also have `t.ToCIDRString()` which guarantees a CIDR format)

**Comment by YairSlobodin1** [comment]
Hi @elazarg 
Could you please also add the following method?
```
func (b *IPBlock) LastIPAddress() string {
	return intToIP4(b.ipRange.Max())
}
```

**Comment by elazarg** [comment]
> Hi @elazarg Could you please also add the following method?
> 
> ```
> func (b *IPBlock) LastIPAddress() string {
> 	return intToIP4(b.ipRange.Max())
> }
> ```

Again - what is the use case?

**Comment by elazarg** [comment]
Isn't it what you get when you convert a set of CIDRs into an ipblock and then take the partitions of `ToCidrList()`?

Regardless of this specific case, much of the point of the models data structures is this king of optimization, so if it's not already supported it should probably be added later, both for specific sets (such as ipblocks) and for products (as in the case of of the triple `IPBlock x Protocol x Port` you describe.

I don't want to add too many features at this point, since already this PR is huge and is a lot of work for @adisos to review. It's better merged first, and updated later.

**Comment by adisos** [comment]
I think the new types should implement basic `String()` functionality, the client code can add other string functionality as needed, but it's also useful for testing, demo, example usage,  debugging, etc.. of this pkg as stand alone .
I would at least implement it for `netset.connecionset`,   `connection.connection`, `product`, `tripleset`..

**Comment by elazarg** [comment]
> I would at least implement it for `netset.connecionset`, `connection.connection`, `product`, `tripleset`

For `Product` and `Tripleset` this means adding String requirement to the interface of `Set`.

**Comment by elazarg** [comment]
> I think the new types should implement basic `String()` functionality, the client code can add other string functionality as needed

And in the case of Product (and TripleSet), the string format should be less specific to the analyzer. I suggest
```
{ (L1 x R1) | ... | (Ln x Rn) } for partitions 1...n
```
So the client - which might not be even aware of this library or its inner working - have an idea what's encoded. Commas can be ambiguous.

File: pkg/netset/connectionset_test.go [hunk 20]
```diff
@@ -122,6 +139,20 @@ func TestConnectionSetBasicOperations(t *testing.T) {
...
+	// dst: 10.240.10.2/31,
+	// conns: protocols 1, src-ports 1-65535, dst-ports 53,protocols 0, src-ports 1-65535, dst-ports 1-65535;all icmp
+	// src: 10.240.10.128/25,
+	// dst: 10.240.10.2/31,
+	// conns: protocols 0, src-ports 1-65535, dst-ports 1-65535;all icmp
+	fmt.Printf("conn12 cubes string:\n%s\n", getPartitionsStr(conn12))
+
+	fmt.Printf("conn17 cubes string:\n%s\n", getPartitionsStr(conn17))
+
 	fmt.Println("done")
```
Comment (elazarg) [inline pkg/netset/connectionset_test.go:156]: This is not a test, it's only printing. Should I transform it into a test?

Comment (adisos) [reply pkg/netset/connectionset_test.go:156]: The relevant `require` checks are already in the test, added this printing for better visibility and documentation of the objects created in the test.

File: pkg/netset/icmpset.go
Comment (zivnevo) [inline pkg/netset/icmpset.go:1]: I have second thoughts about restricting our model to only reason about standard ICMP types and codes.
In both firewalls mechanisms (e.g., NACL) and in actual traffic there is no such restriction.
One may write an NACL rule, denying ICMP traffic with type 20, and although this type is nonstandard, the NACL will do exactly as told: if it observes an ICMP packet with type==20 in its header (possibly tailor-made by an attacker?), the NACL will block this packet.

Comment (elazarg) [reply pkg/netset/icmpset.go:1]: From synthesis POV, I think the case for _allowing_ nonstandard ICMP is likely to be unintentional. It can be denied by default, supported via a special class.

Which makes me think, for both analysis and synthesis, we can have a union of `ICMPset | ICMPInvalid`. That may be better than just handling interval set everywhere in that (a) it's trivial to check and work with well defined ICMP messages and (b) it forces the handling of invalid ICMP messages to be explicit. 

---

Regardless, I think this discussion is better deferred to a dedicated PR (before creating a new release) to make the review feasible. I will gadly work on it, if needed.

Comment (zivnevo) [reply pkg/netset/icmpset.go:1]: Yes, checking if a given ICMP <code,type> pair is standard can be very useful in multiple places (e.g., analyzer's lint).

I prefer defining `StandardICMP` as a subset of the full allowed range (0-254 for type, 0-255 for code).

A separate PR for making this change is a good idea.

**Comment by adisos** [comment]
- Added unit tests for the various types
- Added `String()` method to the relevant interfaces and types 
- opened issue #68  with list of suggested followups after this PR is merged.

**Review by adisos — APPROVED** [review]
